app/cloud.go

```diff
@@ -4,14 +4,123 @@
 package app
 
 import (
+	"encoding/json"
+	"errors"
 	"fmt"
 	"net/http"
+	"sync"
 	"time"
 
+	"github.com/mattermost/mattermost-server/v6/app/request"
 	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/i18n"
 	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+	"github.com/mattermost/mattermost-server/v6/store"
 )
 
+var notifyLock sync.Mutex
+
+func (a *App) NotifySystemAdminsToUpgrade(c *request.Context, currentUserTeamID string) *model.AppError {
+	userId := c.Session().Id
+
+	// check if already notified
+	notifyLock.Lock()
+	sysVal, err := a.Srv().Store.System().GetByName("NOTIFIED_ADMIN_TO_UPGRADE")
+
+	if err != nil {
+		var errNotFound *store.ErrNotFound
+		if !errors.As(err, &errNotFound) {
+			// error was because of something other than the record not existing
+			notifyLock.Unlock()
+			return model.NewAppError("app.NotifySystemAdminsToUpgrade", "api.cloud.notify_admin_to_upgrade_error.fetch", nil, "", http.StatusInternalServerError)
+		}
+		mlog.Debug("NOTIFIED_ADMIN_TO_UPGRADE system value does not exist")
+	}
+
+	alreadyNotifiedUsersInfo := &model.AlreadyCloudNotifiedAdminUsersInfo{
+		Info: make([]model.UserInfo, 0),
+	}
+
+	if sysVal != nil {
+		val := sysVal.Value
+
+		err = json.Unmarshal([]byte(val), alreadyNotifiedUsersInfo)
+		if err != nil {
+			mlog.Error("Unable to Unmarshal", mlog.Err(err))
+		}
+
+		if !alreadyNotifiedUsersInfo.CanNotify(userId) {
+			notifyLock.Unlock()
+			return model.NewAppError("app.NotifySystemAdminsToUpgrade", "api.cloud.notify_admin_to_upgrade_error.already_notified", nil, "", http.StatusForbidden)
+		}
+	}
+
+	team, appErr := a.GetTeam(currentUserTeamID)
+	if appErr != nil {
+		notifyLock.Unlock()
+		return appErr
+	}
+
+	sysadmins, appErr := a.GetUsersFromProfiles(&model.UserGetOptions{
+		Page:     0,
+		PerPage:  100,
+		Role:     model.SystemAdminRoleId,
+		Inactive: false,
+	})
+
+	if appErr != nil {
+		notifyLock.Unlock()
+		return appErr
+	}
+
+	systemBot, appErr := a.GetSystemBot()
+	if appErr != nil {
+		notifyLock.Unlock()
+		return appErr
+	}
+
+	go func() {
+		for _, admin := range sysadmins {
+			T := i18n.GetUserTranslations(admin.Locale)
+			channel, appErr := a.GetOrCreateDirectChannel(c, systemBot.UserId, admin.Id)
+			if appErr != nil {
+				mlog.Error("Error getting direct channel", mlog.Err(appErr))
+				continue
+			}
+
+			post := &model.Post{
+				Message:   T("api.cloud.upgrade_plan_bot_message", map[string]interface{}{"TeamName": team.Name}),
+				UserId:    systemBot.UserId,
+				ChannelId: channel.Id,
+				Type:      fmt.Sprintf("%sup_notification", model.PostCustomTypePrefix), // webapp will have to create renderer for this custom post type
+			}
+
+			_, appErr = a.CreatePost(c, post, channel, false, true)
+			if appErr != nil {
+				mlog.Error("Error creating post", mlog.Err(appErr))
```

Comment:
nit: add more context on what post
```suggestion
				mlog.Error("Error creating post for admin upgrade notification", mlog.Err(appErr))
```